edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/MetadataStore.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Storage;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Newtonsoft.Json;
+
+    public class MetadataStore : IMetadataStore
+    {
+        readonly IKeyValueStore<string, string> metadataEntityStore;
+        readonly string edgeProductInfo;
+
+        public MetadataStore(IKeyValueStore<string, string> metadataEntityStore, string edgeProductInfo)
+        {
+            this.metadataEntityStore = Preconditions.CheckNotNull(metadataEntityStore);
+            this.edgeProductInfo = Preconditions.CheckNotNull(edgeProductInfo, nameof(edgeProductInfo));
+        }
+
+        public async Task<Option<ConnectionMetadata>> GetMetadata(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return await value.Match(
+                async v =>
+                {
+                    return Option.Maybe(await this.GetOrMigrateConnectionMetadata(id, v));
+                },
+                () =>
+                {
+                    return Task.FromResult(Option.None<ConnectionMetadata>());
+                });
+        }
+
+        public async Task<Option<string>> GetModelId(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return value.Match(
+                v =>
+                {
+                    try
+                    {
+                        return JsonConvert.DeserializeObject<ConnectionMetadata>(v).ModelId;
+                    }
+                    catch (JsonException)
+                    {
+                        // If deserialization fails, assume the string is an old productInfo.
+                        return Option.None<string>();
+                    }
+                },
+                () => Option.None<string>());
+        }
+
+        public async Task<string> GetProductInfo(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return await value.Match(
+                async v =>
+                {
+                    return (await this.GetOrMigrateConnectionMetadata(id, v)).ProductInfo;
+                },
+                () => Task.FromResult(string.Empty));
+        }
+
+        async Task<ConnectionMetadata> GetOrMigrateConnectionMetadata(string id, string entityValue)
+        {
+            try
```

Comment:
nit: try catch could be in a different method  TryDeserialize which returns true/false. I think it's more clear than having migration code in the catch function